Implemented INLJ and SMJ

John-Athanasopoulos opened the pull request.

John-Athanasopoulos commented:
@giorgostheo Please review for merge :)

John-Athanasopoulos commented:
@giorgostheo Now joins are in a separate file as per your suggestion

John-Athanasopoulos commented:
Further testing was done after the btree constructor correction and the results were the expected. Thus, I think this branch is ready for review.

Stratis-Dermanoutsos commented:
FYI, the fix was pushed to master.
So, the changes have to be merged from there to develop before this PR is merged.

giorgostheo commented:
Looks good. A couple of comments:

- Right now, the only way inlj or smj are run is if the user specifies so. I like that as a feature, but a natural extension would be to check for an index when a simple join is called, and if an index is present, the call inlj (without the need for explict inlj command). Maybe it is not that interesting/needed, I just wanted to mention it. 
- I would really like to see some tests being implemented. I am think about some joins that join 2 tables from smdb for example and result to a table that we have already save as a csv for example (or as a .sql file, whatever seems easier). Then, if the result is correct (same as the result table that we have saved) the tests pass. This would be very useful since we would have to eventually adapt some sort of testing (proly using pytest). 3-4 join tests should be enough :)

John-Athanasopoulos commented:
Okay perfectly. I will work on both recommendations, so we can either merge this branch and I will fork a new one, or I can keep this one and update it. What do you think is best?

giorgostheo commented:
Keeping this seems the easiest so I suggest we do that, if it's ok with you :)

John-Athanasopoulos commented:
Yes of course, it is easier for me too. I'll update when either of the recommendations is ready

John-Athanasopoulos commented:
Υλοποιούσα το SMJ/INLJ και τώρα που τελείωσε η εξεταστική μπορώ να εστιάσω περισσότερο. Ήθελα να ρωτήσω λοιπόν, επειδή μου είχατε πει η inlj να μην καλείται χειροκίνητα (τύπου select * from X inlj join Y on Z) αλλά μόνο αν υπάρχει index τα εξής 2:
1. Αν δεν έχει index αλλά μπορεί ο ένας από τους 2 πίνακες να γίνει indexed τότε να φτιάχνω temporary index και να τρέχω inlj ή να καλώ κατευθείαν inner join; Συμφέρει δηλαδή για πολυπλοκότητα κατευθείαν να γίνεται inner join ή να φτιάχνεται temporary index και inlj;
2. Να διατηρηθεί η λειτουργικότητα της χειροκίνητης επιλογής τύπου join; Δηλαδή να μπορεί ο χρήστης να πει select * from X smj join Y on Z και αντίστοιχα για inlj; Και αν όχι τότε πότε να καλείται η smj;

John-Athanasopoulos commented:
@giorgostheo  I think this is ready for review :)

giorgostheo commented on miniDB/btree.py:
```diff
@@ -6,11 +6,10 @@ class Node:
         self.b = b # branching factor
-        self.values = values # Values (the data from the pk column)
-        self.ptrs = ptrs # ptrs (the indexes of each datapoint or the index of another bucket)
+        self.values = [] if values is None else values # Values (the data from the pk column)
```
I dont understand why this appears as a change since it is already in the latest branch commit... I dont want you to do anything for this, it is just weird..

giorgostheo commented on mdb.py:
```diff
@@ -123,17 +123,22 @@ def evaluate_from_clause(dic):
         dic['from'] = interpret(subquery)
 
-    join_idx = [i for i,word in enumerate(from_split) if word=='join' and not in_paren(from_split,i)]
+    join_idx = [i for i,word in enumerate(from_split) if (word=='join' or word=='inlj' or word=='smj') and not in_paren(from_split,i)]
```
I understand now why you wanted to keep the syntax as "smj join". This is much more convoluted than it should be. I believe if that syntax is picked ("smj join") then the changes will be minimal here compared to the original.

I think that we should meet in the middle and keep the two word syntax while trimming the last "j" from every join type (example: smj -> sm join or sort_merge join, I dont have an issue with either, pick whichever you like)

giorgostheo commented on miniDB/dashboard.py:
```diff
@@ -1,4 +1,7 @@
-from database import Database
+try:
```
This seems weird.. Maybe use something like sys.path.add (or something like this) instead.

giorgostheo commented on miniDB/externalmergesort.py:
```diff
@@ -0,0 +1,137 @@
+import heapq
```
why is this a separate file and not in joins.py?

giorgostheo commented on miniDB/joins.py:
```diff
@@ -0,0 +1,168 @@
+        self.index = index
+        self.index_saved = index_saved
+
+    def runner(self):
```
runner is not a very descriptive name, maybe swap it for "join" if it doesn't brake anything.

giorgostheo commented on tests/InljSmj_test.py:
```diff
@@ -0,0 +1,182 @@
+# region imports
```
This a comment regarding the way you implemented tests in general... Since testing is something that needs to be added in general, I think that it is very important to have a more or less "stable" initial implementation.

I think that all this should not be needed in order to have a couple of functioning tests. Everything (I think) that you do using python can be done using a .sql (.msql) file. My idea is that:

- test dir has a lot of subdirs that contain .msql files and other files (like .csv)
- each subsubsub...dir contains a single .msql that is run using a small script that is called from the miniDB directory (the head of the dir tree)
- Each .msql file creates all the needed tables, both the ones that will be used to test a query (the initial ones) and the ground truth ones (the ones that correspond to the correct result).
- Lastly, we check if these tables are the same (I dont think this function exists, we should add it...)

Something like this should allow anyone to create tests very quickly and (given the small script), everything should run in one go (and quickly).

John-Athanasopoulos commented:
Hello once again :) Everything revolving the implementation of INLJ and SMJ is done, so this is ready for a final review. Things changed:
1. Changed 'inlj' to 'inl join' (thought that 'index-nested-loop join' was kind of big for a query keyword).
2. Changed the custom imports in every file. In database.py, assigning the value of the Table class in the sys modules was needed, don't know any way around that if it needs to be changed.
3. ExternalMergeSort Class was moved as a subclass of SMJ in joins.py.
4. Runner function name was changed to join. Probably no problems will arise.
5. Tests were removed to be implemented in a more stable way in the following weeks.

With all these changes, the tests (before being removed) were running smoothly once again, at least on my environment (I don't think that there's going to be any problem though).

<img width="960" alt="image" src="https://user-images.githubusercontent.com/61366457/182637963-dde415ab-e7aa-4704-86f4-54367fd9af33.png">

For anything else, please let me know and I will be happy to resolve any issues :)

giorgostheo commented:
Well Done
